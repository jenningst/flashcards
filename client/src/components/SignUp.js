import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link, withRouter } from 'react-router-dom';
import { WaveSpinner } from 'react-spinners-kit';
import { useAuth } from '../contexts/auth-context';
import { createUserWithEmail } from '../contexts/auth-context';
import { PrimaryButton } from './elements/Button';
import { Input } from './elements/Input';
import { Title2, Caption3 } from './elements/Text';
import ROUTE_CONFIG from '../constants/route-config';

const SignUp = ({ history }) => {
  const { registerWithEmail } = useAuth();
  const [email, setEmail] = useState('');
  const [passwordOne, setPasswordOne] = useState('');
  const [passwordTwo, setPasswordTwo] = useState('');
  const [submitting, setSubmitting] = useState(false);
  const [error, setError] = useState(null);
  const handleEmailChange = e => setEmail(e.target.value);
  const handlePasswordOneChange = e => setPasswordOne(e.target.value);
  const handlePasswordTwoChange = e => setPasswordTwo(e.target.value);

  const handleSignUp = async (e) => {
    e.preventDefault();
    // validate form...
    // if invalid:
    // leave email but provide error message if improperly formatted
    // leave passwords but provide error message if not matching

    setSubmitting(true);
    try {
      await createUserWithEmail(email, passwordOne);
      history.push(ROUTE_CONFIG.auth.DASHBOARD);
    } catch (err) {
      switch (err.code) {
        case 'auth/email-already-in-use':
          setError(`It looks like ${email} is already in use. Try using a different email address or logging in.`);
          setEmail('');
          setPasswordOne('');
          setPasswordTwo('');
          setSubmitting(false);
          break;
        case 'auth/weak-password':
          setError(`The password you used is too short. Please use a password with a length of at least 6 characters.`);
          setPasswordOne('');
          setPasswordTwo('');
          setSubmitting(false);
          break;
        default:
          setError(err.message);
          setPasswordOne('');
          setPasswordTwo('');
          setSubmitting(false);
          break;
      }
    }
  };


  // TODO: provide tooltip for password requirements
  // TODO: probably shouldn't be validating their passwords just yet...
  const invalidInputs = 
    passwordOne !== passwordTwo ||
    !passwordOne ||
    email === '';

  return (
    <SignUpPageWrapper className="SignUp">
      <FormWrapper 
        className={`form-wrapper${submitting ? '--submitting' : ''}`}
      >
        <SignUpForm className="SignUp__form">
          <Title2 className="SignUp__title form-title">Sign Up</Title2>
          <ErrorMessageBox className="SignUp__errors">
            <Caption3>{error}</Caption3>
          </ErrorMessageBox>
          <Input
            className="SignUp__input-email"
            aria-label="email"
            type="email"
            id="email"
            onChange={e => handleEmailChange(e)}
            placeholder="Email Address"
            value={email}
          />
          <Input
            className="SignUp__input-password-one"
            aria-label="password-one"
            id="password-one"
            onChange={e => handlePasswordOneChange(e)}
            placeholder="Password"
            value={passwordOne}
          />
          <Input
            className="SignUp__input-password-two"
            aria-label="password-two"
            id="password-two"
            onChange={e => handlePasswordTwoChange(e)}
            placeholder="Confirm Password"
            value={passwordTwo}
          />
          <SignUpButton
            className="SignUp__button-sign-up"
            type="button"
            onClick={e => handleSignUp(e)}
            disabled={invalidInputs}
          >
            Sign Up
          </SignUpButton>
        </SignUpForm>
        <span>
          <Caption3>Already have an account? <Link to={"/login"}>Log in</Link></Caption3>
        </span>
      </FormWrapper>
      <SpinnerWrapper
        className={`loading-spinner-wrapper${submitting ? '--visible' : 'hidden'}`}
      >
        {submitting
          ? <>
              <WaveSpinner
                className="Login__spinner"
                size={30}
                loading={submitting}
                color={'#939CE8'}
              />
              <Title2>Logging In</Title2>
            </>
          : null
        }
      </SpinnerWrapper>
    </SignUpPageWrapper>
  );
};

const SignUpPageWrapper = styled.div`
  box-sizing: border-box;
  display: flex;
  flex-flow: row nowrap;
  justify-content: center;
  align-items: center;
  height: 100%;
`;

const SpinnerWrapper= styled.div`
  position: absolute;
  box-sizing: border-box;
  flex-grow: 1;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;

  opacity: 0.9;

  &[class*="hidden"] {
    z-index: -1;
  }

  &[class*="visible"] {
    z-index: 99999;
  }
`;

const FormWrapper = styled.div`
  box-sizing: border-box;
  flex-grow: 1;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 2rem;
  background: ${props => props.theme.color.main.pureWhite};

  a {
    color: ${props => props.theme.color.main.primary};
    text-decoration: none;
    outline: none;
    border-bottom: 1px solid transparent;

    &:hover {
      transition: color .2s ease-out, border-bottom-color .2s ease-out;
      border-bottom: 1px solid ${props => props.theme.color.main.primaryHover};
    }
  }

  @media(min-width: 376px) {
    min-width: 320px;
    max-width: 450px;
    height: auto;
    box-shadow: 0px 10px 18px -11px rgba(120,119,120,1);
    border-radius: 1rem;
  }

  &[class*="submitting"] {
    opacity: 0.2;
  }
`;

const ErrorMessageBox = styled.span`
  width: 100%;
  height: 2rem;
  color: ${props => props.theme.color.font.danger};
`;

const SignUpForm = styled.form`
  box-sizing: border-box;
  flex-grow: 1;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  width: 100%;

  .form-title {
    width: 100%;
    text-align: left;
    margin-bottom: 1.5rem;
  }

  .ancillary-form-controls {
    display: flex;
    justify-content: space-between;
    width: 100%;
    margin-top: 2rem;
  }
`;

const SignUpButton = styled(PrimaryButton)`
  margin: 1.5rem;
  width: 100%;
`;

SignUp.propTypes = {
  history: PropTypes.object,
}

export default withRouter(SignUp);