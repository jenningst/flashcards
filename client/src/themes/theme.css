/* 

button: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      active: '#5860A5',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    border: {
      primary: '#42487C',
      active: '#939CE8'
      disabled: '#b1aeb7',
      input: '#e9eaf0',
    }, // TODO: consolidate everything below; using the above
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }

}*/