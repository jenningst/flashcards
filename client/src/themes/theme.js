export const darkTheme = {
  font: {
    primary: '#FFFFFF',
    secondary: '#E0E0E0',
    tertiary: '#666666',
  },
  background: {
    primary: '#282626',
    secondary: '#313131',
    highlight: '#6121BF',
    special: '#7358F7',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};

export const lightTheme = {
  color: {
    button: {
      primary: '#939CE8',
      primaryHover: '#7883E2',
      active: '#5860A5',
      secondary: '#50cdff',
      secondaryHover: '#6FD6FF',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#33364d',
      blackCoral: '#585A6D',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1',
      placeholder: '#e9eaf0',
      danger: '#DE6A60',
    },
    border: {
      primary: '#5860A5',
      secondary: '#7883E2',
      disabled: '#b1aeb7',
      inputDefault: '#e9eaf0',
      transparent: '#0000ffff',
      eerieBlack: '#16171D',
      charleston: '#33364d',
    }, // TODO: consolidate everything below; using the above
    background: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
    },
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }
  },
  spacing: {
    compact: '1rem',
    comfortable: '1.5rem',
    generous: '2rem',
  },
  font: {
    primary: '#4e5167',
    secondary: '#262626',
    tertiary: '#666666',
  },
  background: {
    primary: '#ffffff',
    secondary: '#fcfbfc',
    highlight: '#9388ff',

    special: '#50cdff',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};