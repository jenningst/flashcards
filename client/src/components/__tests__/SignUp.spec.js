import React from 'react';
import { StaticRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { darkTheme, lightTheme } from '../../themes/theme';

import {
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react';

import SignUp from '../SignUp';

afterEach(cleanup);

const renderComponent = ({ theme }) =>
  render(
    <ThemeProvider theme={theme}>
      <StaticRouter>
        <SignUp />
      </StaticRouter>
    </ThemeProvider>
  );

describe('<SignUp /> spec', () => {
  it('assert component matches snapshot', () => {
    const { asFragment } = renderComponent({ theme: lightTheme });
    expect(asFragment()).toMatchSnapshot();
  });

  it('assert it initially renders an empty input for email', () => {
    const { getByLabelText } = renderComponent({ theme: lightTheme });
    const emailElement = getByLabelText(/^email address$/i);
    expect(emailElement).toBeInTheDocument();
    expect(emailElement).toBeEmpty();
  });

  it('assert it initially renders an empty input for password one', () => {
    const { getByLabelText } = renderComponent({ theme: lightTheme });
    const passwordElement = getByLabelText(/^password$/i);
    expect(passwordElement).toBeInTheDocument();
    expect(passwordElement).toBeEmpty();
  });

  it('assert it initially renders an empty input for password two', () => {
    const { getByLabelText } = renderComponent({ theme: lightTheme });
    const passwordElement = getByLabelText(/^confirm password$/i);
    expect(passwordElement).toBeInTheDocument();
    expect(passwordElement).toBeEmpty();
  });

  it('assert it initially renders a disabled submit button', () => {
    const { getByText } = renderComponent({ theme: lightTheme });
    expect(getByText(/^sign up$/i)).toBeDisabled();
  });

  it('assert it renders an enabled submit button when all inputs have valid values', () => {
    const { getByLabelText, getByText } = renderComponent({ theme: lightTheme });
    const emailElement = getByLabelText(/^email address$/i);
    const passwordOneElement = getByLabelText(/^password$/i);
    const passwordTwoElement = getByLabelText(/^confirm password$/i);
    const mockEmailEvent = { target: { value: 'test@gmail.com' }};
    const mockPasswordEvent = { target: { value: 'ABC123' }};
    fireEvent.change(emailElement, mockEmailEvent);
    fireEvent.change(passwordOneElement, mockPasswordEvent);
    fireEvent.change(passwordTwoElement, mockPasswordEvent);
    expect(getByText(/^sign up$/i)).toBeEnabled();
  });

  it('assert it renders a link to login page', () => {
    const { getByText } = renderComponent({ theme: lightTheme });
    expect(getByText(/^already have an account/i)).toBeInTheDocument();
  });

  it('assert link redirects to sign up page', () => {
    const { getByText } = renderComponent({ theme: lightTheme });
    expect(getByText(/^already have an account/i)).toHaveAttribute('href', '/login');
  });
});