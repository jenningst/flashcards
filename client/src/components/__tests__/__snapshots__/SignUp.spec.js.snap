// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> spec assert component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="SignUp"
  >
    <label
      for="email"
    >
      Email Address
    </label>
    <input
      class="SignUp__input-email"
      id="email"
      placeholder="Enter Email Address"
      value=""
    />
    <label
      for="password-one"
    >
      Password
    </label>
    <input
      class="SignUp__input-password-one"
      id="password-one"
      placeholder="Enter Password"
      value=""
    />
    <label
      for="password-two"
    >
      Confirm Password
    </label>
    <input
      class="SignUp__input-password-two"
      id="password-two"
      placeholder="Confirm Password"
      value=""
    />
    <button
      class="SignUp__button-sign-up sc-bwzfXH dVyRhp"
      disabled=""
      type="button"
    >
      Sign Up
    </button>
    <a
      href="/login"
    >
      Already have an account? Log in here.
    </a>
  </div>
</DocumentFragment>
`;
